core/src/rpc/basic_context.rs

```diff
@@ -0,0 +1,94 @@
+use std::collections::BTreeMap;
+
+use crate::{
+	dbs::Session,
+	kvs::Datastore,
+	rpc::RpcContext,
+	sql::{Array, Value},
+};
+
+use super::{args::Take, Data, RpcError};
+
+pub struct BasicRpcContext<'a> {
+	pub kvs: &'a Datastore,
+	pub session: Session,
+	pub vars: BTreeMap<String, Value>,
+	pub version_string: String,
+}
+
+impl<'a> BasicRpcContext<'a> {
+	pub fn new(
+		kvs: &'a Datastore,
+		session: Session,
+		vars: BTreeMap<String, Value>,
+		version_string: String,
+	) -> Self {
+		Self {
+			kvs,
+			session,
+			vars,
+			version_string,
+		}
+	}
+}
+
+impl RpcContext for BasicRpcContext<'_> {
+	fn kvs(&self) -> &Datastore {
+		self.kvs
+	}
+
+	fn session(&self) -> &Session {
+		&self.session
+	}
+
+	fn session_mut(&mut self) -> &mut Session {
+		&mut self.session
+	}
+
+	fn vars(&self) -> &BTreeMap<String, Value> {
+		&self.vars
+	}
+
+	fn vars_mut(&mut self) -> &mut BTreeMap<String, Value> {
+		&mut self.vars
+	}
+
+	fn version_data(&self) -> impl Into<super::Data> {
+		Value::Strand(self.version_string.clone().into())
+	}
+
+	// reimplimentaions:
```

Comment:
Don't we need invalidate here? Or does this implementation cover both WS & HTTP

Reply:
The reason for the reimplementations was to optimise them, because a reference and a mutable reference are needed at the same time, the default implementation has to clone session. But as session is a field on these structs this is not necessary.

Reply:
BasicRpcContext is not actually used anywhere, but I've made the same reimplementations on Connection and PostRpcContext